src/goto-harness/recursive_initialization.cpp

```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
   const recursion_sett &known_tags,
   code_blockt &body)
 {
-  body.add(code_assignt{lhs, side_effect_expr_nondett{lhs.type()}});
+  if(
+    lhs.id() != ID_symbol ||
+    !is_array_size_parameter(to_symbol_expr(lhs).get_identifier()))
+  {
+    body.add(code_assignt{lhs, side_effect_expr_nondett{lhs.type()}});
+  }
+}
+
+void recursive_initializationt::initialize_array(
+  const exprt &array,
+  std::size_t depth,
+  const recursion_sett &known_tags,
+  code_blockt &body)
+{
+  PRECONDITION(array.type().id() == ID_array);
+  const auto &array_type = to_array_type(array.type());
```

Comment:
⛏ I think you should change the type of `array` to `array_exprt` - that way you wouldn't have to put these invariants here.

Reply:
I don't think that works here: an `array_exprt` is not the same as an `array_typet`-typed expression. (An `array_exprt` is an array, while an `array_typet`-typed expression could as well just be a symbol.)